packages/@lwc/template-compiler/src/codegen/static-element.ts

```diff
@@ -156,54 +156,48 @@ export function getStaticNodes(root: Root, state: State): Set<ChildNode> {
+// preserveComments = true => [[text], comment, [text]]
+export function transformStaticChildren(elm: StaticElement, preserveComments: boolean) {
     const children = elm.children;
     if (!children.length || !STATIC_ELEMENT_WITH_DYNAMIC_TEXT_SET.has(elm)) {
         // The element either has no children or its children does not contain dynamic text.
         return children;
     }
 
     if (STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.has(elm)) {
+        // This will be hit by serializeStaticElement
         return STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.get(elm)!;
     }
 
     const result: (StaticChildNode | Text[])[] = [];
     const len = children.length;
 
     let current: StaticChildNode;
-    let next: StaticChildNode;
     let contiguousTextNodes: Text[] | null = null;
 
     for (let i = 0; i < len; i++) {
         current = children[i];
-        if (!isText(current)) {
-            contiguousTextNodes = null;
-            result.push(current);
-        } else {
+        if (isText(current)) {
             if (!isNull(contiguousTextNodes)) {
                 // Already in a contiguous text node chain
                 // All contiguous nodes represent an expression in the source, it's guaranteed by the parser.
                 contiguousTextNodes.push(current);
             } else {
-                next = children[i + 1];
-                if (isExpression(current) || (!isUndefined(next) && isText(next))) {
-                    // Text nodes can appear as follows:
-                    // 1. A single text literal node.
-                    // 2. A single text expression node.
-                    // 3. Contiguous series of text nodes (literal/expression mixed) with at least 1 expression.
-                    // When there is an expression in the source, the text nodes are split into contiguous text nodes.
-                    // When there is no expression in the source, the text will appear as a single text literal.
-                    // We normalize all of the contiguous text nodes or single text expression to an array.
-                    // Single text literal nodes (no expression or are not part of a contiguous set of text nodes) remain text nodes
-                    // and will not be consolidated to an array.
-                    // This is to normalize the traversal behavior when creating static parts and when serializing
-                    // the elements.
-                    contiguousTextNodes = [current];
-                }
-                // When contiguousTextNodes is null it is a single string literal.
-                result.push(contiguousTextNodes ?? current);
+                // First time seeing a contiguous text chain
+                contiguousTextNodes = [current];
+                result.push(contiguousTextNodes);
+            }
+        } else {
+            // Non-text nodes signal the end of contiguous text node chain
+            if (!isComment(current) || preserveComments) {
```

Comment:
I'm not following the logic here. I'm reading this as "If the current node is not text and (if it is is not a comment or if we are preserving comments), then we consider this the end of the contiguous text nodes". But that doesn't match up in my head with what I'm seeing within the if block.

Can you help me understand this?

Reply:
> If the current node is not text and (if it is is not a comment or if we are preserving comments), then we consider this the end of the contiguous text nodes

Yup, that's correct. 

The purpose of this function is return the same array of children with the adjacent text nodes grouped into sub-arrays.

By the time we've reached this if block, we've confirmed that the current node is not a text node. 

Then we check the following:

When `preserveComments` is `false`, we add all non-comment nodes directly to the `result` array and set `contiguousTextNodes = null`.

When `preserveComments` is `true`, we add all non-text nodes directly to the `result` array and set `contiguousTextNodes = null`.

Reply:
For example:

```html
{dynamic} text <!-- comment --> text <span></span>
```

The AST nodes will appear like this:

`[text, text, comment, text, element]`

When `preserveComments = true`:

After the first 2 iterations:

`contiguousTextNodes` = [text, text]
`result` = [[text, text]]

After the **third** iteration:

`contiguousTextNodes` = null
`result` = [[text, text], comment]

After the **fourth** iteration:

`contiguousTextNodes` = [text]
`result` = [[text, text], comment, [text]]

After the **fifth** iteration:

`contiguousTextNodes` = null
`result` = [[text, text], comment, [text], element]

Reply:
AST nodes = `[text, text, comment, text, element]`

When `preserveComments = false`:

After the first 2 iterations:

`contiguousTextNodes` = [text, text]
`result` = [[text, text]]

After the third iteration:

`contiguousTextNodes` = [text, text] 
`result` = [[text, text]]

_Note we skipped the comment node_

After the fourth iteration:

`contiguousTextNodes` = [text, text, text]
`result` = [[text, text, text]]

After the fifth iteration:

`contiguousTextNodes` = null
`result` = [[text, text, text], element]

Reply:
Let me know if this helps! 

If it does I'll add some more comments to the function to make it more clear.